### ISSUE TITLE
Add comments to commits

### ISSUE THREAD

**Issue by bradleyfalzon** [bradleyfalzon/gopherci#75]
When implementing checking pushes (not pull requests), I neglected to have GopherCI write comments on those commits.

I don't know why I skipped doing that, I believe this to be a mistake, and GopherCI should post comments whether it was discovered in a PR or a push.

**Comment by shurcooL** [comment]
That explains why it's been surprisingly quiet. :p

**Comment by bradleyfalzon** [comment]
Yeah, sorry, and I'll mark it as a bug, it's a straight forward fix, but just focusing on web component right now.

**Comment by bradleyfalzon** [comment]
OK, planning this now. I'll likely take the opportunity to refactor the comments into a Reporter interface of some description. There will be multiple reporters, such as:

- PR inline comments: needs to dedupe and write only maxIssueComments, this is current behaviour.
- Commit comments: needs to only write maxIssueComments, this allows us to comment on a commit (this issue).
- Status API: Needs to set success or failure.
- PR review comments: needs to dedupe and write only maxIssueComments, using the pull request review comment methods. See #112.
- PR comment summary: Posts a comment on the PR itself with a summary explaining which linters have which number of errors. This probably won't be implemented, but it's just an option.
- Send email: Sends the person submitting the request an email with a summary and link to the analysis.

I won't implement all of these, and some tools may be configured to not report some errors in the future, see #3.

**Comment by bradleyfalzon** [comment]
OK, here's the blocker.

A commit comment occurs at a particular ref, file and diff position: https://developer.github.com/v3/repos/comments/#create-a-commit-comment

If you push a commit and there's an issue, we can easily comment on the exact location.

But if you push two commits, and the issue was in the first commit, GopherCI doesn't have knowledge of that commit ref, it knows ref before and after (head). Further, the diff that we generate is between before and after, so if there's an issue on after/head, the diff position is still likely to be wrong.

There's a few options the immediately come to mind:

1. Skip commenting on a push, and instead send the user our own email. I don't hate this, but it feels like a kludge, the user might not want an email and enjoy to GitHub workflow (I'm looking at you @shurcooL).
1. Run an analysis for every commit. Due to the number of potential commits, and even with rate limiting, this would be too exhaustive.
1. Pass the Analyser a list of all the commits instead of just before and after, have it determine which commit and position an issue occurs on.

Option 3 is basically the only viable option, so I'll be going with that if I don't have any other thoughts that aren't hacks.

**Comment by bradleyfalzon** [comment]
This is going to be too difficult to do.

- GopherCI wants to run the tools just once for all commits.
- We need to know which issue occurred in which commit and the position in that commit.

The problem comes with this example:

```diff
--- /dev/null
+++ b/main.go
@@ -0,0 +1,7 @@
+package main
+
+import "fmt"
+
+func main() {
+       fmt.Println("Hi")
+}
```

This is our initial correct program. There are no problems. We push this code to GitHub.

```diff
--- a/main.go
+++ b/main.go
@@ -3,5 +3,5 @@ package main
 import "fmt"

 func main() {
-       fmt.Println("Hi")
+       fmt.Println("Hi %s", "1") // possible formatting directive in Println call
 }
```

In our next commit, we've introduced a problem: `main.go:6: possible formatting directive in Println call`

We still haven't pushed yet.

```diff
--- a/main.go
+++ b/main.go
@@ -2,6 +2,7 @@ package main

 import "fmt"

+func init() {}
 func main() {
        fmt.Println("Hi %s", "1") // possible formatting directive in Println call
 }
```

We've now added some lines above the problem, committed and pushed.

GopherCI sees the problem as `main.go:7: possible formatting directive in Println call`. But individually neither of the two last patches see line 7 as being modified, only the aggregate patch sees them as being modified:

```diff
--- a/main.go
+++ b/main.go
@@ -2,6 +2,7 @@ package main

 import "fmt"

+func init() {}
 func main() {
-       fmt.Println("Hi")
+       fmt.Println("Hi %s", "1") // possible formatting directive in Println call
 }
```

The aggregate patch is what GopherCI sees right now, and we can record the issues with the aggregate patch position, and because of the way we fetch patches in the analysis summary page, we can even show these issues in line.

But GitHub requires us to know the commit ref and position relative to that commit of where the problem was introduced.

Unless we could somehow merge these patches in reverse order (checking HEAD, then HEAD~1 HEAD, then HEAD~2 HEAD) until we found a modified line at that position, could this work, but may still have false positives, what if a different change occurred on that line in an older commit that wasn't our actual issue.

Another alternative could be that GopherCI does run the tools for every commit (or more practically, the last n commits). In the above example, GopherCI would see the problems for both commits (as it does appear in each commit), so it'd need to deduplicate them somehow, but another commit just before we pushed could remove the issue entirely - this would result in us reporting issues in older commits that aren't issues anymore. This would be correct but an annoyance, yes the problem occurred at that commit (again if we can deduplicate correctly), but it's not a problem when I pushed it, so it's probably nothing to worry about (perhaps it was a style issue).

Having said all of that, right now, GopherCI can show the diff of all commits and exactly which line the issue is on and display it to the user - it just can't comment inline.

This leads me back to two options:

- Leave the behaviour as is, providing an option for GopherCI to email the user if there's a problem.
- Comment on the latest commit, but not on a position and just say there are issues with the push (although there's nothing to suggest to the user these commits were pushed at the same time in GitHub, GopherCI knows though), and link the user to GopherCI.

**Comment by bradleyfalzon** [comment]
Example pushing multiple commits, with many commits introducing issues (9 in total).

![screenshot 2017-08-10 14 25 29](https://user-images.githubusercontent.com/1834577/29154980-d4489654-7dd7-11e7-85f1-56af9b4bf23f.png)

If we comment each issue on the commit itself (not defining path or position in https://developer.github.com/v3/repos/comments/#create-a-commit-comment), it looks like so:

![screenshot 2017-08-10 14 27 30](https://user-images.githubusercontent.com/1834577/29155031-36e9b0a4-7dd8-11e7-8ddb-30346fab17b1.png)

This can easily be improved with (potentially adding filename and line number as a link to the correct location/branch etc):

![screenshot 2017-08-10 14 33 36](https://user-images.githubusercontent.com/1834577/29155131-f8b03ef6-7dd8-11e7-9cc4-b54822040854.png)

Or more simply:

![screenshot 2017-08-10 14 39 06](https://user-images.githubusercontent.com/1834577/29155238-c51940b4-7dd9-11e7-99d9-8edfb8b1f72b.png)
